core/arch/arm/pta/benchmark.c

```diff
@@ -62,43 +69,61 @@ static TEE_Result register_benchmark_memref(uint32_t type,
 {
 	TEE_Result res;
 
-	if ((TEE_PARAM_TYPE_GET(type, 0) != TEE_PARAM_TYPE_MEMREF_INOUT) ||
+	if ((TEE_PARAM_TYPE_GET(type, 0) != TEE_PARAM_TYPE_VALUE_INOUT) ||
 		(TEE_PARAM_TYPE_GET(type, 1) != TEE_PARAM_TYPE_NONE) ||
 		(TEE_PARAM_TYPE_GET(type, 2) != TEE_PARAM_TYPE_NONE) ||
 		(TEE_PARAM_TYPE_GET(type, 3) != TEE_PARAM_TYPE_NONE)) {
 		return TEE_ERROR_BAD_PARAMETERS;
 	}
 
-	/*
-	 * We accept only non-secure buffers, as we later perform
-	 * registration of this buffer in NS layers
-	 * (optee linux kmod/optee client)
-	 */
-	if (!tee_vbuf_is_non_sec(p[0].memref.buffer, p[0].memref.size))
-		return TEE_ERROR_BAD_PARAMETERS;
-
 	mutex_lock(&bench_reg_mu);
 
 	/* Check if we have already registered buffer */
 	if (bench_ts_global) {
-		EMSG("Timestamp buffer was already registered\n");
+		EMSG(TA_PRINT_PREFIX
+			"timestamp buffer was already registered\n");
+		mutex_unlock(&bench_reg_mu);
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	bench_ts_size = p[0].value.b;
+	if (!bench_ts_size) {
+		EMSG(TA_PRINT_PREFIX
+			"timestamp buffer size isn't specified\n");
+		mutex_unlock(&bench_reg_mu);
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	bench_mobj = thread_rpc_alloc_payload(bench_ts_size, &bench_c);
+	if (!bench_mobj) {
+		EMSG(TA_PRINT_PREFIX
+			"can't create mobj for timestamp buffer\n");
+		mutex_unlock(&bench_reg_mu);
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	bench_ts_global = (struct tee_ts_global *)
+		mobj_get_va(bench_mobj, 0);
+	if (!bench_ts_global) {
+		thread_rpc_free_payload(bench_c, bench_mobj);
+
 		mutex_unlock(&bench_reg_mu);
 		return TEE_ERROR_BAD_STATE;
 	}
 
-	DMSG("Registering timestamp buffer, addr = %p, paddr = %" PRIxPA "\n",
-			p[0].memref.buffer,
-			virt_to_phys(p[0].memref.buffer));
-	bench_ts_global = p[0].memref.buffer;
+	DMSG(TA_PRINT_PREFIX
+		"allocated timestamp buffer, addr = %p\n",
+		(void *)bench_ts_global);
 
 	mutex_unlock(&bench_reg_mu);
 
 	/* Send back to the optee linux kernel module */
 	res = rpc_reg_global_buf(OPTEE_MSG_RPC_CMD_BENCH_REG_NEW,
 			virt_to_phys((void *)bench_ts_global),
-			sizeof(struct tee_ts_global) +
-			sizeof(struct tee_ts_cpu_buf) *
-			bench_ts_global->cores);
+			bench_ts_size);
+
+	p[0].value.a = virt_to_phys(bench_ts_global);
```

Comment:
case 64bit phys address?

Reply:
seems same issue was already in `get_benchmark_memref()`. Maybe to be fixed is a specific P-R.

Reply:
I agree, probably cleaner to deal with that in a separate PR.

Reply:
@etienne-lms thanks, will fix